File: InterleavingString.cpp
/*
 * Time Limit Exceeded
*/
#include<iostream>
#include<string>

using namespace std;

class Solution {
public:
    bool isInterleave(string s1, string s2, string s3) {
        
        return isInterleave(0, 0, 0, 1, s1, s2, s3)||isInterleave(0, 0, 0, 2, s1, s2, s3);
        
    }
    
    bool isInterleave(int a, int b, int c, int type, string s1, string s2, string s3){
        string t;
        int x;
        if(type==1){
            t=s1;
            x=a;
        }else{
            t=s2;
            x=b;
        }
        
        if(a==s1.length()&&b==s2.length()&&c==s3.length()){
            return true;
        }
        
        for(int i=x;i<t.length();i++){
            if(t[i]==s3[c+i-x]){
                if((type==1&&isInterleave(i+1, b, c+i-x+1, 2, s1, s2, s3))||(type==2&&isInterleave(a, i+1, c+i-x+1, 1, s1, s2, s3))){
                    return true;
                }
            }else{
                return false;
            }
        }
        
        return false;
    }
};

int main(){
    string s1, s2, s3;
    cin>>s1>>s2>>s3;
    
    Solution *solution=new Solution();
    solution->isInterleave(s1, s2, s3) ? cout<<"True" : cout<<"False";
    
    return 0;
}

File: InterleavingString.py
class Solution:
    def isInterleave(self, s1, s2, s3):
        """
        :type s1: str
        :type s2: str
        :type s3: str
        :rtype: bool
        """
        if len(s1)+len(s2)!=len(s3):
            return False
        dp = [False]*(len(s2)+1)
        dp[0] = True
        for i in range(len(s1)+1):
            for j in range(len(s2)+1):
                if i==0 and j==0:
                    continue
                if (j>0 and dp[j-1] and s2[j-1]==s3[i+j-1]) or (i>0 and dp[j] and s1[i-1]==s3[i+j-1]):
                    dp[j] = True
                else:
                    dp[j] = False
                
        return dp[len(s2)]
        
if __name__=='__main__':
    s1 = raw_input()
    s2 = raw_input()
    s3 = raw_input()
    
    solution = Solution()
    print(solution.isInterleave(s1, s2, s3))
